### ISSUE TITLE
CLI client

### ISSUE THREAD

**Issue by titanous** [flynn-archive/strowger#11]
The existing client example should be a full-fledged CLI client that allows access to 100% of the RPC interface. (started in #4)

**Comment by miracle2k** [comment]
So I would like to address this next.
1. The RPC interface

Currently, `AddFrontend` takes a `Config` struct, which used to have Type=HTTP + Domain + Service. It now has the TLS config as well, in who knows, maybe more in the future. To implement `RemoveFrontend`, and also something necessary as `UpdateFrontend`, I think these frontends need to be keyed on something.

Since internally, we are already keying on the domain, it seems to make sense to expose this externally in the same way.

That raises the question if the current API and terminology should be maintained. That is, does it make sense to have an API `UpdateFrontend(config)`, where `config.Domain` will be used to setup the internal mapping, or could we just as well use one of these:

```
 UpdateFrontend(domain, config)
 UpdateDomain(name, domainCfg)
 UpdateRoute(domainMatch, config)
```

Maybe you can add some notes as to what your ideas for this originally where.
1. I might be inclined to work on the terminology because at times, I find it confusing. `Server` is used for the main object and "httpServer" represents a backend service, `HTTPFrontend` also sort of is the server as a whole, but "Frontend" is the public name for a domain. 
2. The command line client currently calls the RPC API directly. Should it use the "client library" in `/client/client.go`?

**Comment by titanous** [comment]
> Since internally, we are already keying on the domain, it seems to make sense to expose this externally in the same way.

Yeah, we also need to figure out what the key is for TCP frontends (port?).

> I might be inclined to work on the terminology because at times, I find it confusing. Server is used for the main object and "httpServer" represents a backend service, HTTPFrontend also sort of is the server as a whole, but "Frontend" is the public name for a domain. 

I'm not sold at all on the current terminology, feel free to make suggestions and I'll rethink this soon. Maybe "routes" instead of "frontends"?

> The command line client currently calls the RPC API directly. Should it use the "client library"?

Yep!

**Comment by miracle2k** [comment]
TCP currently seems as if designed to auto-allocate the ports. I would have assumed I can pick the port. I would have also assumed I can serve two separate Jabber servers on 5222 via two separate IPs - I only realize now that this is not straightforward.

---

Currently the classes look like this:

```
Router -> Server  (are both the same currently)
    HTTPFrontend
        domain
        httpServer
    TCPFrontend
        tcpServer
```

Which really is fine; I'd only:
- find a different name for domains added to a frontend on the API side. "routes" makes more sense to me if more complex path-based routing is planned; otherwise, I'd just go with "domain".
- might rename "httpServer" to "httpBackend".

---

As for the CLI itself, should we use different subcommands for HTTP and TCP modes?

```
./strowger-cli http add domain service --opts
./strowger-cli http set domain
./strowger-cli http list
./strowger-cli tcp list
```

What about building the CLI into the ./strowger binary itself?

**Comment by titanous** [comment]
We should add an option so that the port can be picked, the TCP implementation is incomplete. In the future we may support pools of IPs so that the same port may be used by different applications, but this gets tricky.

---

I like `HTTPRoute` and `TCPRoute`.

---

Yeah, subcommands are fine, but I'm okay with (`http-add`, `tcp-list`, etc) too. The CLI should be a separate package/binary (`strowgerctl`).

**Comment by miracle2k** [comment]
I thought I would just work on this, but I believe it needs some more discussion after all.

My problem was: How to model an API to update the certs for a domain (but not other fields).
- UpdateHTTPRouteTLSConfig(domain, cert, byte)  ?
- UpdateHTTPRoute(domain, configMap[string]string) - the map containing only the config options that should change?
- Or should the API simply require the whole configuration of a route to be sent, i.e. a client might do `route = GetHTTPRoute() && route.cert = foo && UpdateHTTPRoute(route)`
- Maybe something I am missing...

But really, `etcdctl set` is already seductively simple. We need to think about what value we can add on top of that at what cost.

Right now, the code flow for API interactions is this:

strowgerctl -> strowger client library -> strowger RPC server -> internal AddDomain -> write to etcd -> read from etcd

Adding a new option means introducing it _at every single step_. That is a lot of boilerplate for creating one or two keys in etcd.

**Comment by titanous** [comment]
Instead of using individual keys in etcd, we should probably serialize a config struct to/from JSON and store it in a single key. I'm fine with requiring the entire config to be sent when modifying a route.

**Comment by titanous** [comment]
@miracle2k I think I've addressed most of the nomenclature stuff in 59881540f0652c86039d81af3351af617504eea6.

**Comment by titanous** [comment]
@miracle2k I switched to storing the config as JSON (814718454a55d93afda5c09b7533dce281365ccc).
